python/nav/ipdevpoll/plugins/typeoid.py

```diff
@@ -22,133 +21,147 @@
-                                 oldname, newname)
-            self._logger.debug("old=%r new=%r", self.netbox.type, type_)
-
-            if not type_:
-                return self.create_new_type()
-        return type_
+                sender=self, netbox_id=self.netbox.id, new_type=new_type
+            )
 
     @defer.inlineCallbacks
-    def create_new_type(self):
-        """Creates a new NetboxType from the collected sysObjectID."""
-        vendor_id = yield db.run_in_thread(get_vendor_id, self.sysobjectid)
-        vendor = self.containers.factory(vendor_id, shadows.Vendor)
-        vendor.id = vendor_id
-
-        type_ = self.containers.factory(self.sysobjectid, shadows.NetboxType)
-        type_.vendor = vendor
-        type_.name = self.sysobjectid
-        type_.sysobjectid = self.sysobjectid
-
-        def _set_sysdescr(descr):
-            self._logger.debug("Creating new type with descr=%r", descr)
-            type_.description = descr
+    def _create_new_type(self, oid):
+        """Creates a new NetboxType from the given sysobjectid."""
+        self._logger.debug("Creating a new type from %r", oid)
+        description = yield self.snmpv2_mib.get_sysDescr()
+
+        def _create():
+            vendor = self._get_vendor(oid)
+            type_ = manage.NetboxType(
+                vendor=vendor,
+                name=str(oid),
+                sysobjectid=str(oid).strip("."),
+                description=description,
+            )
+            type_.save()
             return type_
 
-        yield self.snmpv2_mib.get_sysDescr().addCallback(_set_sysdescr)
-
-#
-# Helper functions
-#
-
-
-def get_vendor_id(sysobjectid):
-    """Looks up the most likely vendorid based on a sysObjectID"""
-    enterprise = get_enterprise_id(sysobjectid)
-    cx = connection.cursor()
-    cx.execute("SELECT vendorid FROM enterprise_number "
-               "WHERE enterprise = %s LIMIT 1", (enterprise,))
-    vendorid = cx.fetchone()
-    return vendorid[0] if vendorid else make_new_vendor_id(sysobjectid)
-
-
-def make_new_vendor_id(sysobjectid):
-    """Makes up a new vendorid based on a sysObjectID"""
-    enterprise = get_enterprise_id(sysobjectid)
-    if enterprise in _enterprise_map:
-        name = _enterprise_map[enterprise]
-        name = name.replace(CONSTANT_PREFIX, '').replace('_', '').lower()[:15]
-        return name
+        new_type = yield db.run_in_thread(_create)
+        defer.returnValue(new_type)
+
+    @classmethod
+    def _get_vendor(cls, sysobjectid):
+        """Looks up the most likely vendor based on a sysObjectID"""
+        enterprise = get_enterprise_id(sysobjectid)
+        query = (
+            "vendorid IN (SELECT vendorid FROM enterprise_number WHERE enterprise=%s)"
+        )
+        try:
+            return manage.Vendor.objects.extra(where=[query], params=[enterprise])[0]
+        except IndexError:
+            return cls._make_new_vendor(sysobjectid)
+
```

Comment:
Ah, yes, I was wondering why `connection` was imported in the old code, and it was was as unnecessary as I thought.